src/crypto/chacha20poly1305/chacha20.go

```diff
@@ -0,0 +1,169 @@
+package chacha20poly1305
+
+import (
+	"crypto/cipher"
+	"strconv"
+	"unsafe"
+)
+
+type chacha20Cipher struct {
+	state [16]uint32
+	buf   [16]uint32
+	avail int
+}
+
+type KeySizeError int
+type IVSizeError int
+
+const outputBufferSize = 64
+
+func (k KeySizeError) Error() string {
+	return "crypto/chacha20poly1305: invalid key size " + strconv.Itoa(int(k))
+}
+
+func (i IVSizeError) Error() string {
+	return "crypto/chacha20poly1305: invalid iv size " + strconv.Itoa(int(i))
+}
+
+func u8tou32(in []byte) uint32 {
+	if supportsUnaligned {
+		return *(*uint32)(unsafe.Pointer(&in[0]))
+	} else {
+		return uint32(in[0]) ^ uint32(in[1])<<8 ^ uint32(in[2])<<16 ^ uint32(in[3])<<24
+	}
+}
+
+func newChacha(key []byte) (*chacha20Cipher, error) {
+	k := len(key)
+
+	if k != 32 {
+		return nil, KeySizeError(k)
+	}
+
+	c := chacha20Cipher{state: [16]uint32{0x61707865, 0x3320646e, 0x79622d32, 0x6b206574,
+		u8tou32(key[0:4]), u8tou32(key[4:8]), u8tou32(key[8:12]), u8tou32(key[12:16]),
+		u8tou32(key[16:20]), u8tou32(key[20:24]), u8tou32(key[24:28]), u8tou32(key[28:32]),
+		0, 0, 0, 0},
+		avail: 0}
+
+	return &c, nil
+}
+
+func (c *chacha20Cipher) setIV(iv []byte, ctr uint32) error {
+	i := len(iv)
+
+	if i != 12 {
+		return IVSizeError(i)
+	}
+
+	c.state[12] = ctr
+	c.state[13] = u8tou32(iv[0:4])
+	c.state[14] = u8tou32(iv[4:8])
+	c.state[15] = u8tou32(iv[8:12])
+	c.avail = 0
+
+	return nil
+}
+
+func NewCipher(key, iv []byte, ctr uint32) (cipher.Stream, error) {
```

Comment:
This should return a Cipher object instead of a interface, so that users have a concrete type to type asset to, for example.